File: src/CSharp/SodaFlow.Tests/ForwardReferenceTests.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using TUnit.Assertions;
using TUnit.Assertions.Enums;
using TUnit.Assertions.Extensions;
using TUnit.Core;

namespace SodaFlow.Tests;

public sealed class ForwardReferenceTests
{
    private sealed class Child(Cell<Node> parent)
    {
        public Cell<Node> Parent { get; } = parent;
    }

    private sealed class Node(Child child)
    {
        public Child Child { get; } = child;

        public static Node WithChildHolding(Cell<Node> reference) => new(new Child(reference));
    }

    [Test]
    public async Task TestWithoutCapturesResolvesTheReference()
    {
        Node node =
            ForwardReference<Node>.WithoutCaptures(static reference => Node.WithChildHolding(reference.AsCell()));

        await Assert.That(node.Child.Parent.Sample()).IsSameReferenceAs(node);
    }

    [Test]
    public async Task TestWithoutCapturesReturnsWhatTheFunctionProduced()
    {
        object? produced = null;

        object result =
            ForwardReference<object>.WithoutCaptures(_ =>
            {
                produced = new object();
                return produced;
            });

        await Assert.That(result).IsSameReferenceAs(produced);
    }

    [Test]
    public async Task TestWithoutCapturesRunsTheFunctionOnce()
    {
        int calls = 0;

        // ReSharper disable once ReturnValueOfPureMethodIsNotUsed - Testing for side effect only.
        ForwardReference<int>.WithoutCaptures(_ =>
        {
            calls++;
            return 1;
        });

        await Assert.That(calls).IsEqualTo(1);
    }

    [Test]
    public async Task TestWithoutCapturesReferenceNeverChanges()
    {
        // The single-valued case of a cell loop: the reference resolves once and stays there.
        Node node =
            ForwardReference<Node>.WithoutCaptures(static reference => Node.WithChildHolding(reference.AsCell()));

        List<Node> @out = [];

        using (node.Child.Parent.ListenStrong(@out.Add))
        {
        }

        await Assert.That(@out).IsEquivalentTo([node], CollectionOrdering.Matching);
    }

    [Test]
    public async Task TestWithCapturesResolvesTheReferenceAndReturnsTheCaptures()
    {
        (Node node, StreamSink<int> sink) =
            ForwardReference<Node>.WithCaptures(static reference =>
                (Value: Node.WithChildHolding(reference.AsCell()), Captures: Stream.CreateSink<int>()));

        await Assert.That(node.Child.Parent.Sample()).IsSameReferenceAs(node);
        await Assert.That(sink).IsNotNull();
    }

    [Test]
    public async Task TestWithCapturesRunsTheFunctionOnce()
    {
        int calls = 0;

        // ReSharper disable once ReturnValueOfPureMethodIsNotUsed - Testing for side effect only.
        ForwardReference<int>.WithCaptures(_ =>
        {
            calls++;
            return (Value: 1, Captures: 2);
        });

        await Assert.That(calls).IsEqualTo(1);
    }

    [Test]
    public async Task TestTwoObjectsCanReferToEachOther()
    {
        // Neither exists when the other is constructed, which is the knot this unties.
        (Node node, Child child) =
            ForwardReference<Node>.WithCaptures(static reference =>
            {
                Child c = new(reference.AsCell());
                return (Value: Node.WithChildHolding(reference.AsCell()), Captures: c);
            });

        await Assert.That(child.Parent.Sample()).IsSameReferenceAs(node);
        await Assert.That(node.Child.Parent.Sample()).IsSameReferenceAs(node);
    }

    [Test]
    public async Task TestWorksInsideAnExistingTransaction()
    {
        Node node =
            Transaction.Run(static () =>
                ForwardReference<Node>.WithoutCaptures(static reference => Node.WithChildHolding(reference.AsCell())));

        await Assert.That(node.Child.Parent.Sample()).IsSameReferenceAs(node);
    }

    [Test]
    public async Task TestReferenceCannotBeReadDuringConstruction() =>
        // The reference is a promise about what the value will be, not the value, so asking
        // for it before the constructing function has returned has no answer.
        await Assert.That(static () =>
            // ReSharper disable once ReturnValueOfPureMethodIsNotUsed - Testing for side effect only.
            ForwardReference<int>.WithoutCaptures(static reference => reference.Sample())).ThrowsExactly<InvalidOperationException>();
}

File: src/FSharp/SodaFlow.FSharp.Tests/ForwardReferenceTests.fs
module SodaFlow.Tests.ForwardReference

open System
open System.Collections.Generic
open SodaFlow
open TUnit.Core

type Node(child: Child) =
    member _.Child = child

and Child(parent: Cell<Node>) =
    member _.Parent = parent

/// Builds the node whose child holds a reference back to it, which is the knot under test.
let private nodeHolding reference = Node(Child reference)

type ``Forward Reference Tests``() =

    [<Test>]
    member _.``Test Create With No Captures Resolves The Reference``() =
        task {
            let node = forwardReferenceWithNoCaptures nodeHolding

            do! Expect.Same(node, node.Child.Parent |> sampleC)
        }

    [<Test>]
    member _.``Test Create With No Captures Returns What The Function Produced``() =
        task {
            let produced = obj ()

            let result = forwardReferenceWithNoCaptures (fun _ -> produced)

            do! Expect.Same(produced, result)
        }

    [<Test>]
    member _.``Test Create With No Captures Runs The Function Once``() =
        task {
            let mutable calls = 0

            forwardReferenceWithNoCaptures (fun _ ->
                calls <- calls + 1
                1)
            |> ignore

            do! Expect.Equal(1, calls)
        }

    [<Test>]
    member _.``Test Create With No Captures Reference Never Changes``() =
        task {
            // The single-valued case of a cell loop: the reference resolves once and stays there.
            let node = forwardReferenceWithNoCaptures nodeHolding
            let out = List<_>()
            let l = node.Child.Parent |> listenStrongC out.Add
            l |> unlistenL

            do! Expect.Sequence([ node ], out)
        }

    [<Test>]
    member _.``Test Create Resolves The Reference And Returns The Captures``() =
        task {
            let struct (node, sink) =
                forwardReference (fun reference -> struct (nodeHolding reference, sinkS ()))

            do! Expect.Same(node, node.Child.Parent |> sampleC)
            do! Expect.NotNull(sink)
        }

    [<Test>]
    member _.``Test Create Runs The Function Once``() =
        task {
            let mutable calls = 0

            forwardReference (fun _ ->
                calls <- calls + 1
                struct (1, 2))
            |> ignore

            do! Expect.Equal(1, calls)
        }

    [<Test>]
    member _.``Test Two Objects Can Refer To Each Other``() =
        task {
            // Neither exists when the other is constructed, which is the knot this unties.
            let struct (node, child) =
                forwardReference (fun reference -> struct (nodeHolding reference, Child reference))

            do! Expect.Same(node, child.Parent |> sampleC)
            do! Expect.Same(node, node.Child.Parent |> sampleC)
        }

    [<Test>]
    member _.``Test Works Inside An Existing Transaction``() =
        task {
            let node = runT (fun () -> forwardReferenceWithNoCaptures nodeHolding)

            do! Expect.Same(node, node.Child.Parent |> sampleC)
        }

    [<Test>]
    member _.``Test Reference Cannot Be Read During Construction``() =
        task {
            // The reference is a promise about what the value will be, not the value, so asking for
            // it before the constructing function has returned has no answer.
            do! Expect.Throws<InvalidOperationException>(fun () -> forwardReferenceWithNoCaptures sampleC |> ignore)
        }
